{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>q</th>\n",
       "      <th>TIL-others</th>\n",
       "      <th>#cpg</th>\n",
       "      <th>p(MWU)</th>\n",
       "      <th>p(2dks)</th>\n",
       "      <th>TIL</th>\n",
       "      <th>others</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>Gene/Repeat type</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1034254</td>\n",
       "      <td>1034587</td>\n",
       "      <td>0.016244</td>\n",
       "      <td>-0.149632</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.33591</td>\n",
       "      <td>0.48554</td>\n",
       "      <td>0.321658</td>\n",
       "      <td>0.335913</td>\n",
       "      <td>0.813317</td>\n",
       "      <td>20</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1034555</td>\n",
       "      <td>1034703</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1034254</td>\n",
       "      <td>1034587</td>\n",
       "      <td>0.016244</td>\n",
       "      <td>-0.149632</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.33591</td>\n",
       "      <td>0.48554</td>\n",
       "      <td>0.321658</td>\n",
       "      <td>0.335913</td>\n",
       "      <td>0.813317</td>\n",
       "      <td>20</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1034555</td>\n",
       "      <td>1034703</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1034254</td>\n",
       "      <td>1034587</td>\n",
       "      <td>0.016244</td>\n",
       "      <td>-0.149632</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.33591</td>\n",
       "      <td>0.48554</td>\n",
       "      <td>0.321658</td>\n",
       "      <td>0.335913</td>\n",
       "      <td>0.813317</td>\n",
       "      <td>20</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1034555</td>\n",
       "      <td>1034703</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1034254</td>\n",
       "      <td>1034587</td>\n",
       "      <td>0.016244</td>\n",
       "      <td>-0.149632</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.33591</td>\n",
       "      <td>0.48554</td>\n",
       "      <td>0.321658</td>\n",
       "      <td>0.335913</td>\n",
       "      <td>0.813317</td>\n",
       "      <td>20</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1034555</td>\n",
       "      <td>1034703</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1034254</td>\n",
       "      <td>1034587</td>\n",
       "      <td>0.016244</td>\n",
       "      <td>-0.149632</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.33591</td>\n",
       "      <td>0.48554</td>\n",
       "      <td>0.321658</td>\n",
       "      <td>0.335913</td>\n",
       "      <td>0.813317</td>\n",
       "      <td>20</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1034555</td>\n",
       "      <td>1034703</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end         q  TIL-others  #cpg    p(MWU)   p(2dks)  \\\n",
       "0  chr1  1034254  1034587  0.016244   -0.149632    20  0.000012  0.000066   \n",
       "1  chr1  1034254  1034587  0.016244   -0.149632    20  0.000012  0.000066   \n",
       "2  chr1  1034254  1034587  0.016244   -0.149632    20  0.000012  0.000066   \n",
       "3  chr1  1034254  1034587  0.016244   -0.149632    20  0.000012  0.000066   \n",
       "4  chr1  1034254  1034587  0.016244   -0.149632    20  0.000012  0.000066   \n",
       "\n",
       "       TIL   others        10        11        12  13    14       15       16  \\\n",
       "0  0.33591  0.48554  0.321658  0.335913  0.813317  20  chr1  1034555  1034703   \n",
       "1  0.33591  0.48554  0.321658  0.335913  0.813317  20  chr1  1034555  1034703   \n",
       "2  0.33591  0.48554  0.321658  0.335913  0.813317  20  chr1  1034555  1034703   \n",
       "3  0.33591  0.48554  0.321658  0.335913  0.813317  20  chr1  1034555  1034703   \n",
       "4  0.33591  0.48554  0.321658  0.335913  0.813317  20  chr1  1034555  1034703   \n",
       "\n",
       "  Gene/Repeat type Region  \n",
       "0             AGRN    cds  \n",
       "1             AGRN    cds  \n",
       "2             AGRN    cds  \n",
       "3             AGRN    cds  \n",
       "4             AGRN    cds  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "ingenefile=\"/Users/irffanalahi/Research/Research_update/SM/melcfdref/bestworkingref/genomic_analysis/pbltilmel_mincp2outcutoff_addedcol_genomic_feature_withrepeat_header_celltypeseperated/TIL_row_combined.txt\"\n",
    "outfile=ingenefile+\"_prioritizedregion.txt\"\n",
    "priority_list=['promoter','cds','3utrexon','5utrexon','intron','repeat']#['repeat','cds','3utrexon','5utrexon','promoter','intron']#['promoter','cds','3utrexon','5utrexon','intron','repeat']\n",
    "\n",
    "ingenedf=pd.read_csv(ingenefile,sep=\"\\t\")\n",
    "\n",
    "ingenedf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped=ingenedf.groupby(['chrom','start','end'])\n",
    "rowlist=[]\n",
    "for name, group in grouped:\n",
    "    for region in priority_list:\n",
    "        tmp= group[group['Region']==region]\n",
    "        if tmp.shape[0]>0:\n",
    "            tmprow=(tmp.iloc[0]).copy()\n",
    "            \n",
    "            if (tmprow['Region']=='3utrexon') | (tmprow['Region']=='5utrexon'):\n",
    "                tmprow['Region']='3/5 utrexon'\n",
    "                \n",
    "            rowlist.append(tmprow)  \n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdf=pd.DataFrame(rowlist)\n",
    "outdf=outdf[['chrom','start','end','Region']]\n",
    "duplicateRowsDF = outdf[outdf.duplicated()]\n",
    "if duplicateRowsDF.shape[0]>0:\n",
    "    print(\"duplicated rows.WHY???? FIX it before plotting\")\n",
    "    print(duplicateRowsDF)\n",
    "outdf.to_csv(outfile,sep=\"\\t\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
