{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>q</th>\n",
       "      <th>diff</th>\n",
       "      <th>#cpg</th>\n",
       "      <th>p(MWU)</th>\n",
       "      <th>p(2dks)</th>\n",
       "      <th>CD8TIL</th>\n",
       "      <th>others</th>\n",
       "      <th>...</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>Gene/Repeat type</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>973577</td>\n",
       "      <td>973714</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>-0.509535</td>\n",
       "      <td>6</td>\n",
       "      <td>7.765300e-11</td>\n",
       "      <td>5.848000e-09</td>\n",
       "      <td>0.3295</td>\n",
       "      <td>0.83903</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552111</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.335889</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.596556</td>\n",
       "      <td>chr1</td>\n",
       "      <td>973499</td>\n",
       "      <td>973640</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>973577</td>\n",
       "      <td>973714</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>-0.509535</td>\n",
       "      <td>6</td>\n",
       "      <td>7.765300e-11</td>\n",
       "      <td>5.848000e-09</td>\n",
       "      <td>0.3295</td>\n",
       "      <td>0.83903</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552111</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.335889</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.596556</td>\n",
       "      <td>chr1</td>\n",
       "      <td>973499</td>\n",
       "      <td>973640</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>973577</td>\n",
       "      <td>973714</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>-0.509535</td>\n",
       "      <td>6</td>\n",
       "      <td>7.765300e-11</td>\n",
       "      <td>5.848000e-09</td>\n",
       "      <td>0.3295</td>\n",
       "      <td>0.83903</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552111</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.335889</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.596556</td>\n",
       "      <td>chr1</td>\n",
       "      <td>973499</td>\n",
       "      <td>973640</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>973577</td>\n",
       "      <td>973714</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>-0.509535</td>\n",
       "      <td>6</td>\n",
       "      <td>7.765300e-11</td>\n",
       "      <td>5.848000e-09</td>\n",
       "      <td>0.3295</td>\n",
       "      <td>0.83903</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552111</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.335889</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.596556</td>\n",
       "      <td>chr1</td>\n",
       "      <td>973499</td>\n",
       "      <td>973640</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>973577</td>\n",
       "      <td>973714</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>-0.509535</td>\n",
       "      <td>6</td>\n",
       "      <td>7.765300e-11</td>\n",
       "      <td>5.848000e-09</td>\n",
       "      <td>0.3295</td>\n",
       "      <td>0.83903</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552111</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.335889</td>\n",
       "      <td>-0.322917</td>\n",
       "      <td>-0.596556</td>\n",
       "      <td>chr1</td>\n",
       "      <td>973499</td>\n",
       "      <td>973640</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end         q      diff  #cpg        p(MWU)       p(2dks)  \\\n",
       "0  chr1  973577  973714  0.000005 -0.509535     6  7.765300e-11  5.848000e-09   \n",
       "1  chr1  973577  973714  0.000005 -0.509535     6  7.765300e-11  5.848000e-09   \n",
       "2  chr1  973577  973714  0.000005 -0.509535     6  7.765300e-11  5.848000e-09   \n",
       "3  chr1  973577  973714  0.000005 -0.509535     6  7.765300e-11  5.848000e-09   \n",
       "4  chr1  973577  973714  0.000005 -0.509535     6  7.765300e-11  5.848000e-09   \n",
       "\n",
       "   CD8TIL   others  ...        29        30        31        32        33  \\\n",
       "0  0.3295  0.83903  ... -0.552111 -0.322917 -0.335889 -0.322917 -0.596556   \n",
       "1  0.3295  0.83903  ... -0.552111 -0.322917 -0.335889 -0.322917 -0.596556   \n",
       "2  0.3295  0.83903  ... -0.552111 -0.322917 -0.335889 -0.322917 -0.596556   \n",
       "3  0.3295  0.83903  ... -0.552111 -0.322917 -0.335889 -0.322917 -0.596556   \n",
       "4  0.3295  0.83903  ... -0.552111 -0.322917 -0.335889 -0.322917 -0.596556   \n",
       "\n",
       "     34      35      36  Gene/Repeat type  Region  \n",
       "0  chr1  973499  973640           PLEKHN1     cds  \n",
       "1  chr1  973499  973640           PLEKHN1     cds  \n",
       "2  chr1  973499  973640           PLEKHN1     cds  \n",
       "3  chr1  973499  973640           PLEKHN1     cds  \n",
       "4  chr1  973499  973640           PLEKHN1     cds  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "ingenefile=\"/Users/irffanalahi/Research/Research_update/SM/melcfdref/LargerLTME/BluOurPBL/Dashboard/BlOurPBLourCD8TIL_out_mincpg2_q0.5_addedcol_q0.00001_diff0.1_hypo_noheader_genomic_feature_withrepeat_header_celltypeseperated/rowcombined/CD8TIL_row_combined.txt\"\n",
    "outfile=ingenefile+\"_prioritizedregion.txt\"\n",
    "priority_list=['promoter','cds','3utrexon','5utrexon','intron','repeat']\n",
    "\n",
    "ingenedf=pd.read_csv(ingenefile,sep=\"\\t\")\n",
    "\n",
    "ingenedf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped=ingenedf.groupby(['chrom','start','end'])\n",
    "rowlist=[]\n",
    "for name, group in grouped:\n",
    "    for region in priority_list:\n",
    "        tmp= group[group['Region']==region]\n",
    "        if tmp.shape[0]>0:\n",
    "            tmprow=(tmp.iloc[0]).copy()\n",
    "            \n",
    "            if (tmprow['Region']=='3utrexon') | (tmprow['Region']=='5utrexon'):\n",
    "                tmprow['Region']='3/5 utrexon'\n",
    "                \n",
    "            rowlist.append(tmprow)  \n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdf=pd.DataFrame(rowlist)\n",
    "outdf=outdf[['chrom','start','end','Region']]\n",
    "duplicateRowsDF = outdf[outdf.duplicated()]\n",
    "if duplicateRowsDF.shape[0]>0:\n",
    "    print(\"duplicated rows.WHY???? FIX it before plotting\")\n",
    "    print(duplicateRowsDF)\n",
    "outdf.to_csv(outfile,sep=\"\\t\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
